File: backend/services/source_editor.py
from __future__ import annotations

import json
import re
from html import escape
from datetime import datetime
from pathlib import Path
from typing import Iterable

import pdfplumber
from docx import Document
from pptx import Presentation
from reportlab.lib.pagesizes import letter
from reportlab.lib.styles import getSampleStyleSheet
from reportlab.platypus import PageBreak, Paragraph, SimpleDocTemplate, Spacer
from sqlalchemy import select
from sqlalchemy.orm import Session

from .. import models
from ..config import settings
from ..llm import LLMResult
from ..rag import retriever
from .adaptation import _resolve_provider

ALLOWED_EXTENSIONS = {".docx", ".pptx", ".pdf"}
EDIT_DIR = settings.uploads_dir / "lesson_edits"
EDIT_DIR.mkdir(parents=True, exist_ok=True)


def _norm(text: str) -> set[str]:
    return {t for t in re.findall(r"[a-z0-9]+", text.lower()) if len(t) > 2}


def _safe_filename(name: str) -> str:
    cleaned = re.sub(r"[^A-Za-z0-9._-]+", "_", name).strip("._")
    return cleaned or "lesson_edit"


def _resolve_source_path(source_path: str) -> Path:
    base = settings.sample_lessons_dir.resolve()
    candidate = (base / source_path).resolve()
    if base not in candidate.parents and candidate != base:
        raise LookupError("source file must be inside Sample Lessons")
    if not candidate.exists() or candidate.suffix.lower() not in ALLOWED_EXTENSIONS:
        raise LookupError("source file not found or unsupported")
    return candidate


def source_files_for_lesson(lesson: models.Lesson) -> list[dict]:
    lesson_tokens = _norm(lesson.title)
    files: list[dict] = []
    for path in settings.sample_lessons_dir.rglob("*"):
        if not path.is_file() or path.suffix.lower() not in ALLOWED_EXTENSIONS:
            continue
        rel = path.relative_to(settings.sample_lessons_dir)
        rel_tokens = _norm(str(rel))
        score = len(lesson_tokens & rel_tokens)
        if score == 0:
            continue
        files.append(
            {
                "source_path": rel.as_posix(),
                "filename": path.name,
                "file_type": path.suffix.lower().lstrip("."),
                "size_bytes": path.stat().st_size,
                "_score": score,
            }
        )
    files.sort(key=lambda f: (-f["_score"], f["filename"].lower()))
    for item in files:
        item.pop("_score", None)
    return files


def _docx_blocks(path: Path) -> tuple[Document, list[tuple[str, object]]]:
    doc = Document(str(path))
    blocks: list[tuple[str, object]] = []
    for p in doc.paragraphs:
        if p.text.strip():
            blocks.append((p.text, p))
    for table in doc.tables:
        for row in table.rows:
            for cell in row.cells:
                for p in cell.paragraphs:
                    if p.text.strip():
                        blocks.append((p.text, p))
    return doc, blocks


def _pptx_blocks(path: Path) -> tuple[Presentation, list[tuple[str, object]]]:
    prs = Presentation(str(path))
    blocks: list[tuple[str, object]] = []
    for slide in prs.slides:
        for shape in slide.shapes:
            if getattr(shape, "has_text_frame", False) and shape.text.strip():
                blocks.append((shape.text, shape))
    return prs, blocks


def _pdf_blocks(path: Path) -> list[tuple[str, object]]:
    blocks: list[tuple[str, object]] = []
    with pdfplumber.open(str(path)) as pdf:
        for i, page in enumerate(pdf.pages, start=1):
            text = page.extract_text() or ""
            if text.strip():
                blocks.append((text.strip(), {"page": i}))
    return blocks


def _split_text_for_lengths(text: str, lengths: list[int]) -> list[str]:
    if not lengths:
        return []
    parts: list[str] = []
    pos = 0
    for i, length in enumerate(lengths):
        if i == len(lengths) - 1:
            parts.append(text[pos:])
        else:
            parts.append(text[pos : pos + length])
            pos += length
    return parts


def _set_docx_paragraph_text_preserving_runs(paragraph, new_text: str) -> None:
    # Editing the existing w:t nodes preserves paragraph style, run formatting,
    # and hyperlink XML containers far better than assigning paragraph.text.
    text_nodes = [
        node
        for node in paragraph._p.iter()
        if node.tag == "{http://schemas.openxmlformats.org/wordprocessingml/2006/main}t"
    ]
    if not text_nodes:
        paragraph.text = new_text
        return
    lengths = [len(node.text or "") for node in text_nodes]
    if not any(lengths):
        lengths = [0 for _ in text_nodes]
        lengths[0] = len(new_text)
    for node, part in zip(text_nodes, _split_text_for_lengths(new_text, lengths)):
        node.text = part


def _set_pptx_shape_text_preserving_runs(shape, new_text: str) -> None:
    runs = []
    for paragraph in shape.text_frame.paragraphs:
        runs.extend(paragraph.runs)
    if not runs:
        shape.text = new_text
        return
    lengths = [len(run.text or "") for run in runs]
    if not any(lengths):
        lengths = [0 for _ in runs]
        lengths[0] = len(new_text)
    for run, part in zip(runs, _split_text_for_lengths(new_text, lengths)):
        run.text = part


def _chunks(items: list[tuple[int, str]], max_chars: int = 7000) -> Iterable[list[tuple[int, str]]]:
    chunk: list[tuple[int, str]] = []
    size = 0
    for item_id, text in items:
        item_size = len(text)
        if chunk and size + item_size > max_chars:
            yield chunk
            chunk = []
            size = 0
        chunk.append((item_id, text))
        size += item_size
    if chunk:
        yield chunk


_JSON_FENCE = re.compile(r"```(?:json)?\s*(\{.*?\})\s*```", re.DOTALL)


def _parse_items(text: str) -> dict[int, str]:
    candidate = text.strip()
    match = _JSON_FENCE.search(candidate)
    if match:
        candidate = match.group(1)
    else:
        first = candidate.find("{")
        last = candidate.rfind("}")
        if first != -1 and last != -1:
            candidate = candidate[first : last + 1]
    data = json.loads(candidate)
    return {int(item["id"]): str(item["text"]) for item in data.get("items", [])}


def _kb_context(db: Session, lesson: models.Lesson, cluster_id: int | None, kb_ids: list[int]) -> str:
    if not kb_ids:
        return ""
    kb_rows = list(
        db.execute(select(models.KnowledgeBase).where(models.KnowledgeBase.kb_id.in_(kb_ids))).scalars()
    )
    specs = [{"kb_id": kb.kb_id, "kb_name": kb.kb_name, "category": kb.category} for kb in kb_rows]
    cluster = db.get(models.StudentCluster, cluster_id) if cluster_id else None
    query = " ".join(
        filter(
            None,
            [
                lesson.title,
                lesson.cs_topic,
                lesson.objectives,
                cluster.cluster_name if cluster else None,
                cluster.cluster_description if cluster else None,
            ],
        )
    )
    chunks = retriever.retrieve_for_lesson(query, specs, top_k_per_kb=2)
    lines: list[str] = []
    if cluster:
        lines.append(f"Learner profile: {cluster.cluster_name} - {cluster.cluster_description or ''}")
    for chunk in chunks:
        lines.append(f"[KB #{chunk.kb_id} {chunk.kb_name} - {chunk.section_title}]\n{chunk.text}")
    return "\n\n".join(lines)


def _rewrite_texts(
    db: Session,
    *,
    teacher: models.Teacher,
    lesson: models.Lesson,
    source_filename: str,
    instruction: str,
    texts: list[str],
    cluster_id: int | None,
    kb_ids: list[int],
) -> tuple[list[str], LLMResult]:
    provider = _resolve_provider(db, teacher.teacher_id)
    indexed = [(i, text) for i, text in enumerate(texts)]
    rewritten = list(texts)
    last_result: LLMResult | None = None
    kb_context = _kb_context(db, lesson, cluster_id, kb_ids)
    system = (
        "You edit teacher lesson source files. Preserve meaning, classroom usefulness, numbering, "
        "placeholders, URLs, hyperlink labels, and references unless the teacher explicitly asks to change them. "
        "Follow the teacher instruction. Return JSON only."
    )
    for batch in _chunks(indexed):
        user = {
            "lesson": {
                "title": lesson.title,
                "grade_level": lesson.grade_level,
                "cs_topic": lesson.cs_topic,
                "objectives": lesson.objectives,
            },
            "source_file": source_filename,
            "teacher_instruction": instruction,
            "rag_context": kb_context,
            "items": [{"id": item_id, "text": text} for item_id, text in batch],
            "required_output": {
                "items": [
                    {
                        "id": "same numeric id from input",
                        "text": "edited replacement text only; do not add explanations",
                    }
                ]
            },
        }
        result = provider.generate(system=system, user=json.dumps(user, ensure_ascii=False), max_tokens=8192)
        last_result = result
        try:
            parsed = _parse_items(result.text)
        except Exception as e:
            raise RuntimeError("LLM returned an unreadable file-edit response") from e
        for item_id, new_text in parsed.items():
            if 0 <= item_id < len(rewritten) and new_text.strip():
                rewritten[item_id] = new_text.strip()
    if last_result is None:
        raise RuntimeError("source file did not contain editable text")
    return rewritten, last_result


def _write_pdf(out_path: Path, title: str, blocks: list[str]) -> None:
    styles = getSampleStyleSheet()
    doc = SimpleDocTemplate(str(out_path), pagesize=letter, title=title)
    story = [Paragraph(title, styles["Title"]), Spacer(1, 12)]
    for i, text in enumerate(blocks):
        if i:
            story.append(PageBreak())
        for para in text.splitlines():
            if para.strip():
                story.append(Paragraph(escape(para.strip()), styles["BodyText"]))
                story.append(Spacer(1, 6))
    doc.build(story)


def edit_source_file(
    db: Session,
    *,
    teacher: models.Teacher,
    lesson_id: int,
    source_path: str,
    instruction: str,
    cluster_id: int | None,
    kb_ids: list[int],
) -> dict:
    lesson = db.get(models.Lesson, lesson_id)
    if not lesson:
        raise LookupError("lesson not found")
    source = _resolve_source_path(source_path)
    rel = source.relative_to(settings.sample_lessons_dir).as_posix()
    if rel not in {f["source_path"] for f in source_files_for_lesson(lesson)}:
        raise LookupError("source file is not linked to this lesson")

    suffix = source.suffix.lower()
    stamp = datetime.now().strftime("%Y%m%d-%H%M%S")
    out_name = _safe_filename(f"{source.stem}_ai_edit_{stamp}{suffix}")
    out_path = EDIT_DIR / out_name

    if suffix == ".docx":
        doc, blocks = _docx_blocks(source)
        texts, targets = [b[0] for b in blocks], [b[1] for b in blocks]
        rewritten, _result = _rewrite_texts(
            db,
            teacher=teacher,
            lesson=lesson,
            source_filename=source.name,
            instruction=instruction,
            texts=texts,
            cluster_id=cluster_id,
            kb_ids=kb_ids,
        )
        for target, new_text in zip(targets, rewritten):
            _set_docx_paragraph_text_preserving_runs(target, new_text)
        doc.save(str(out_path))
    elif suffix == ".pptx":
        prs, blocks = _pptx_blocks(source)
        texts, targets = [b[0] for b in blocks], [b[1] for b in blocks]
        rewritten, _result = _rewrite_texts(
            db,
            teacher=teacher,
            lesson=lesson,
            source_filename=source.name,
            instruction=instruction,
            texts=texts,
            cluster_id=cluster_id,
            kb_ids=kb_ids,
        )
        for target, new_text in zip(targets, rewritten):
            _set_pptx_shape_text_preserving_runs(target, new_text)
        prs.save(str(out_path))
    elif suffix == ".pdf":
        blocks = _pdf_blocks(source)
        texts = [b[0] for b in blocks]
        rewritten, _result = _rewrite_texts(
            db,
            teacher=teacher,
            lesson=lesson,
            source_filename=source.name,
            instruction=instruction,
            texts=texts,
            cluster_id=cluster_id,
            kb_ids=kb_ids,
        )
        _write_pdf(out_path, f"{lesson.title} - AI Edited Copy", rewritten)
    else:
        raise LookupError("unsupported source file type")

    return {
        "filename": out_name,
        "file_type": suffix.lstrip("."),
        "download_url": f"/api/lesson-file-edits/{out_name}",
        "note": (
            "Original source file was not changed. DOCX/PPTX copies preserve existing text runs, hyperlinks, and basic document structure where possible; "
            "PDF copies are regenerated as text-based PDFs rather than exact-layout edits."
        ),
    }


def edited_file_path(filename: str) -> Path:
    safe = Path(filename).name
    path = (EDIT_DIR / safe).resolve()
    if EDIT_DIR.resolve() not in path.parents or not path.exists():
        raise LookupError("edited file not found")
    return path

File: server/src/services/source-editor.js
const fs = require("fs");
const path = require("path");
const mammoth = require("mammoth");
const pdfParse = require("pdf-parse");
const officeparser = require("officeparser");
const { Document, Packer, Paragraph, TextRun, ExternalHyperlink, ImageRun } = require("docx");
const pptxgenjs = require("pptxgenjs");
const { PDFDocument, StandardFonts, rgb } = require("pdf-lib");
const db = require("../db");
const { resolveProvider } = require("./adaptation");
const { retrieveForLesson } = require("./rag/retriever");

const ALLOWED_EXTENSIONS = new Set([".docx", ".pptx", ".pdf"]);
const EDIT_DIR = path.resolve(__dirname, "..", "..", "..", "uploads", "lesson_edits");
fs.mkdirSync(EDIT_DIR, { recursive: true });

function safeFilename(name) {
  const cleaned = name.replace(/[^A-Za-z0-9._-]+/g, "_").replace(/^[._]+|[._]+$/g, "");
  return cleaned || "lesson_edit";
}

function resolveSourcePath(sourcePath) {
  const base = path.resolve(__dirname, "..", "..", "..", "Sample Lessons");
  const candidate = path.resolve(base, sourcePath);
  if (!candidate.startsWith(base)) throw new Error("source file must be inside Sample Lessons");
  if (!fs.existsSync(candidate)) throw new Error("source file not found");
  const ext = path.extname(candidate).toLowerCase();
  if (!ALLOWED_EXTENSIONS.has(ext)) throw new Error("unsupported file type");
  return candidate;
}

async function extractDocxSegments(filePath) {
  const result = await mammoth.convertToHtml({ path: filePath }, {
    convertImage: mammoth.images.dataUri,
  });
  return parseDocxHtml(result.value);
}

function parseDocxHtml(html) {
  const segments = [];
  html = html.replace(/<html[^>]*>|<\/html>|<body[^>]*>|<\/body>/gi, "");
  html = html.replace(/<br\s*\/?>/gi, "\n");

  // Strip table wrappers but keep cell content (Word often uses tables for layout)
  html = html.replace(/<\/?(table|thead|tbody|tfoot|tr)[^>]*>/gi, "");
  html = html.replace(/<t[dh][^>]*>([\s\S]*?)<\/t[dh]>/gi, "$1");

  const regex = /<(p|h[1-6]|li)(?:\s+[^>]*)?>([\s\S]*?)<\/\1>/gi;
  let match;
  while ((match = regex.exec(html)) !== null) {
    const tag = match[1].toLowerCase();
    let content = match[2];

    if (tag === "li") {
      content = "• " + content;
    }

    // Split content by inline images
    const parts = [];
    const imgRegex = /(<img[^>]+src="([^"]+)"[^>]*>)/gi;
    let lastIndex = 0;
    let imgMatch;

    while ((imgMatch = imgRegex.exec(content)) !== null) {
      if (imgMatch.index > lastIndex) {
        const text = htmlToMarkdownText(content.slice(lastIndex, imgMatch.index));
        if (text.trim()) {
          parts.push({ type: "text", content: text.trim() });
        }
      }
      const src = imgMatch[2];
      if (src.startsWith("data:")) {
        const parsed = parseInlineImage(src);
        if (parsed) parts.push({ type: "image", mimeType: parsed.mimeType, base64: parsed.base64, imageType: parsed.type });
      }
      lastIndex = imgRegex.lastIndex;
    }

    if (lastIndex < content.length) {
      const text = htmlToMarkdownText(content.slice(lastIndex));
      if (text.trim()) {
        parts.push({ type: "text", content: text.trim() });
      }
    }

    if (parts.length === 0) {
      const text = htmlToMarkdownText(content);
      if (text.trim()) {
        parts.push({ type: "text", content: text.trim() });
      }
    }

    segments.push(...parts);
  }

  return segments;
}

function htmlToMarkdownText(html) {
  let text = html.replace(/<a\s+href="([^"]+)"[^>]*>([\s\S]*?)<\/a>/gi, (m, url, linkText) => {
    const cleanLinkText = linkText.replace(/<[^>]+>/g, "").trim();
    return `[${cleanLinkText}](${url})`;
  });
  text = text.replace(/<[^>]+>/g, "");
  text = text
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&amp;/g, "&")
    .replace(/&quot;/g, '"')
    .replace(/&apos;/g, "'")
    .replace(/&nbsp;/g, " ");
  return text;
}

function parseInlineImage(src) {
  const match = src.match(/^data:([^;]+);base64,(.+)$/);
  if (!match) return null;
  const mimeType = match[1];
  const base64 = match[2];
  let type = null;
  if (mimeType === "image/png") type = "png";
  else if (mimeType === "image/jpeg" || mimeType === "image/jpg") type = "jpg";
  else if (mimeType === "image/gif") type = "gif";
  else if (mimeType === "image/bmp") type = "bmp";
  else return null;
  return { mimeType, base64, type };
}

function getImageDimensions(buffer) {
  if (buffer.length > 24 && buffer[0] === 0x89 && buffer.toString("ascii", 1, 4) === "PNG") {
    const width = buffer.readUInt32BE(16);
    const height = buffer.readUInt32BE(20);
    return { width, height };
  }
  if (buffer.length > 2 && buffer[0] === 0xff && buffer[1] === 0xd8) {
    let i = 2;
    while (i < buffer.length) {
      if (buffer[i] !== 0xff) {
        i++;
        continue;
      }
      const marker = buffer[i + 1];
      if (marker === 0xd9) break;
      if (marker === 0xd8) {
        i += 2;
        continue;
      }
      if (marker === 0xc0 || marker === 0xc2) {
        const height = buffer.readUInt16BE(i + 5);
        const width = buffer.readUInt16BE(i + 7);
        return { width, height };
      }
      const len = buffer.readUInt16BE(i + 2);
      i += 2 + len;
    }
  }
  if (buffer.length > 10 && buffer.toString("ascii", 0, 6) === "GIF89a") {
    const width = buffer.readUInt16LE(6);
    const height = buffer.readUInt16LE(8);
    return { width, height };
  }
  return null;
}

function scaleDimensions(width, height, maxDim = 500) {
  if (!width || !height) return { width: 400, height: 300 };
  const ratio = Math.min(maxDim / width, maxDim / height);
  if (ratio < 1) {
    return { width: Math.round(width * ratio), height: Math.round(height * ratio) };
  }
  return { width, height };
}

function textToDocxRuns(text) {
  const runs = [];
  const regex = /\[([^\]]+)\]\(([^)]+)\)/g;
  let lastIndex = 0;
  let match;

  while ((match = regex.exec(text)) !== null) {
    if (match.index > lastIndex) {
      runs.push(new TextRun(text.slice(lastIndex, match.index)));
    }
    runs.push(
      new ExternalHyperlink({
        children: [new TextRun({ text: match[1], style: "Hyperlink" })],
        link: match[2],
      })
    );
    lastIndex = regex.lastIndex;
  }

  if (lastIndex < text.length) {
    runs.push(new TextRun(text.slice(lastIndex)));
  }

  if (runs.length === 0) {
    runs.push(new TextRun(text));
  }

  return runs;
}

async function extractPptx(filePath) {
  return new Promise((resolve, reject) => {
    try {
      officeparser.parseOffice(filePath, (result) => {
        if (!result || typeof result.toText !== 'function') {
          return reject(new Error('PPTX extraction failed'));
        }
        resolve(result.toText().split('\n').filter(line => line.trim()));
      });
    } catch (e) {
      reject(e);
    }
  });
}

async function extractPdf(filePath) {
  const buffer = fs.readFileSync(filePath);
  const data = await pdfParse(buffer);
  return data.text.split('\n').filter(line => line.trim());
}

function splitTextForLengths(text, lengths) {
  if (!lengths || lengths.length === 0) return [];
  const parts = [];
  let pos = 0;
  for (let i = 0; i < lengths.length; i++) {
    if (i === lengths.length - 1) {
      parts.push(text.slice(pos));
    } else {
      parts.push(text.slice(pos, pos + lengths[i]));
      pos += lengths[i];
    }
  }
  return parts;
}

function chunks(items, maxChars = 7000) {
  const result = [];
  let chunk = [];
  let size = 0;
  for (const [id, text] of items) {
    const itemSize = text.length;
    if (chunk.length > 0 && size + itemSize > maxChars) {
      result.push(chunk);
      chunk = [];
      size = 0;
    }
    chunk.push([id, text]);
    size += itemSize;
  }
  if (chunk.length > 0) result.push(chunk);
  return result;
}

const JSON_FENCE = /```(?:json)?\s*(\{.*?\})\s*```/s;

function parseItems(text) {
  let candidate = text.trim();
  const match = JSON_FENCE.exec(candidate);
  if (match) {
    candidate = match[1];
  } else {
    const first = candidate.indexOf("{");
    const last = candidate.lastIndexOf("}");
    if (first !== -1 && last !== -1) {
      candidate = candidate.slice(first, last + 1);
    }
  }
  const data = JSON.parse(candidate);
  const result = {};
  for (const item of data.items || []) {
    result[Number(item.id)] = String(item.text);
  }
  return result;
}

async function rewriteTexts({ teacherId, lesson, sourceFilename, instruction, texts, clusterId, kbIds }) {
  const provider = resolveProvider(teacherId);
  const indexed = texts.map((text, i) => [i, text]);
  const rewritten = [...texts];
  let lastResult = null;

  let kbContext = "";
  if (kbIds && kbIds.length > 0) {
    const kbRows = db.prepare(
      "SELECT kb_id, kb_name, category FROM knowledge_base WHERE kb_id IN (" + kbIds.map(() => "?").join(",") + ")"
    ).all(...kbIds);
    const specs = kbRows.map(r => ({ kb_id: r.kb_id, kb_name: r.kb_name, category: r.category }));
    const query = [lesson.title, lesson.cs_topic, lesson.objectives].filter(Boolean).join(" ");
    const retrievedChunks = await retrieveForLesson(query, specs, 2);
    const lines = [];
    for (const c of retrievedChunks) {
      lines.push(`[KB #${c.kb_id} ${c.kb_name} - ${c.section_title}]\n${c.text}`);
    }
    kbContext = lines.join("\n\n");
  }

  const system = (
    "You edit teacher lesson source files. Preserve meaning, classroom usefulness, numbering, " +
    "placeholders, URLs, hyperlink labels, and references unless the teacher explicitly asks to change them. " +
    "Follow the teacher instruction. Return JSON only."
  );

  for (const batch of chunks(indexed)) {
    const user = JSON.stringify({
      lesson: {
        title: lesson.title,
        grade_level: lesson.grade_level,
        cs_topic: lesson.cs_topic,
        objectives: lesson.objectives,
      },
      source_file: sourceFilename,
      teacher_instruction: instruction,
      rag_context: kbContext,
      items: batch.map(([id, text]) => ({ id, text })),
      required_output: {
        items: [
          { id: "same numeric id from input", text: "edited replacement text only; do not add explanations" },
        ],
      },
    });

    const result = await provider.generate({ system, user, maxTokens: 8192 });
    lastResult = result;

    try {
      const parsed = parseItems(result.text);
      for (const [itemId, newText] of Object.entries(parsed)) {
        const idx = Number(itemId);
        if (idx >= 0 && idx < rewritten.length && newText.trim()) {
          rewritten[idx] = newText.trim();
        }
      }
    } catch (e) {
      throw new Error("LLM returned an unreadable file-edit response");
    }
  }

  if (!lastResult) throw new Error("source file did not contain editable text");
  return { rewritten, result: lastResult };
}

async function writeDocx(outPath, segments) {
  const children = [];

  for (const seg of segments) {
    if (seg.type === "text") {
      const runs = textToDocxRuns(seg.content);
      children.push(new Paragraph({ children: runs }));
    } else if (seg.type === "image") {
      const buffer = Buffer.from(seg.base64, "base64");
      const dims = getImageDimensions(buffer);
      const scaled = scaleDimensions(dims?.width, dims?.height);
      const imageRun = new ImageRun({
        type: seg.imageType,
        data: buffer,
        transformation: scaled,
      });
      children.push(new Paragraph({ children: [imageRun] }));
    }
  }

  const doc = new Document({
    sections: [{ children }],
  });
  const buffer = await Packer.toBuffer(doc);
  fs.writeFileSync(outPath, buffer);
}

async function writePptx(outPath, blocks) {
  const pres = new pptxgenjs();
  let slide = pres.addSlide();
  let y = 0.5;
  for (const text of blocks) {
    if (y > 6.5) {
      slide = pres.addSlide();
      y = 0.5;
    }
    slide.addText(text, { x: 0.5, y, w: 9, h: 0.5, fontSize: 14 });
    y += 0.6;
  }
  await pres.writeFile({ fileName: outPath });
}

async function writePdf(outPath, title, blocks) {
  const pdfDoc = await PDFDocument.create();
  const font = await pdfDoc.embedFont(StandardFonts.Helvetica);
  const titleFont = await pdfDoc.embedFont(StandardFonts.HelveticaBold);

  let page = pdfDoc.addPage([612, 792]); // Letter size
  const { height } = page.getSize();
  let y = height - 72;

  page.drawText(title, { x: 72, y, font: titleFont, size: 18, color: rgb(0, 0, 0) });
  y -= 36;

  for (const block of blocks) {
    const lines = block.split('\n').filter(l => l.trim());
    for (const line of lines) {
      if (y < 72) {
        page = pdfDoc.addPage([612, 792]);
        y = height - 72;
      }
      page.drawText(line.trim(), { x: 72, y, font, size: 11, color: rgb(0, 0, 0) });
      y -= 16;
    }
    y -= 8;
  }

  const pdfBytes = await pdfDoc.save();
  fs.writeFileSync(outPath, pdfBytes);
}

async function editSourceFile({ teacherId, lessonId, sourcePath, instruction, clusterId, kbIds }) {
  const lesson = db.prepare("SELECT * FROM lesson WHERE lesson_id = ?").get(lessonId);
  if (!lesson) throw new Error("lesson not found");

  const source = resolveSourcePath(sourcePath);
  const suffix = path.extname(source).toLowerCase();
  const stamp = new Date().toISOString().replace(/[:.]/g, "-").slice(0, 19);
  const baseName = path.basename(source, suffix);
  const outName = safeFilename(`${baseName}_ai_edit_${stamp}${suffix}`);
  const outPath = path.join(EDIT_DIR, outName);

  let texts;
  let docxSegments = null;

  if (suffix === ".docx") {
    docxSegments = await extractDocxSegments(source);
    texts = docxSegments.filter((s) => s.type === "text").map((s) => s.content);
  } else if (suffix === ".pptx") {
    texts = await extractPptx(source);
  } else if (suffix === ".pdf") {
    texts = await extractPdf(source);
  } else {
    throw new Error("unsupported source file type");
  }

  const { rewritten } = await rewriteTexts({
    teacherId, lesson, sourceFilename: path.basename(source),
    instruction, texts, clusterId, kbIds,
  });

  if (suffix === ".docx") {
    const textSegments = docxSegments.filter((s) => s.type === "text");
    for (let i = 0; i < textSegments.length && i < rewritten.length; i++) {
      textSegments[i].content = rewritten[i].trim();
    }
    await writeDocx(outPath, docxSegments);
  } else if (suffix === ".pptx") {
    await writePptx(outPath, rewritten);
  } else if (suffix === ".pdf") {
    await writePdf(outPath, `${lesson.title} - AI Edited Copy`, rewritten);
  }

  return {
    filename: outName,
    file_type: suffix.slice(1),
    download_url: `/api/lesson-file-edits/${outName}`,
    note: "Original source file was not changed. Edited copy preserves hyperlinks and images from the original. Other layout and styling may vary.",
  };
}

function editedFilePath(filename) {
  const safe = path.basename(filename);
  const filePath = path.resolve(EDIT_DIR, safe);
  if (!filePath.startsWith(EDIT_DIR)) throw new Error("invalid filename");
  if (!fs.existsSync(filePath)) throw new Error("edited file not found");
  return filePath;
}

function sourceFilesForLesson(lesson) {
  const lessonTokens = new Set(
    lesson.title.toLowerCase().match(/[a-z0-9]+/g)?.filter(t => t.length > 2) || []
  );
  const base = path.resolve(__dirname, "..", "..", "..", "Sample Lessons");
  if (!fs.existsSync(base)) return [];

  const files = [];
  const walk = (dir) => {
    for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        walk(fullPath);
      } else if (ALLOWED_EXTENSIONS.has(path.extname(entry.name).toLowerCase())) {
        const rel = path.relative(base, fullPath);
        const relTokens = new Set(
          rel.toLowerCase().match(/[a-z0-9]+/g)?.filter(t => t.length > 2) || []
        );
        const score = [...lessonTokens].filter(t => relTokens.has(t)).length;
        if (score > 0) {
          files.push({
            source_path: rel.replace(/\\/g, "/"),
            filename: entry.name,
            file_type: path.extname(entry.name).slice(1),
            size_bytes: fs.statSync(fullPath).size,
          });
        }
      }
    }
  };
  walk(base);
  files.sort((a, b) => b.filename.localeCompare(a.filename));
  return files;
}

module.exports = { editSourceFile, editedFilePath, sourceFilesForLesson };
